### ISSUE TITLE
MySQL error on `bin/run build_bookworm` in main-notraefik (a dumb question issue)

### ISSUE THREAD

**Issue by senderle** [senderle/bookworm-compose#14]
I decided to strip out the extra `traefik` config from `main` because I think it's unnecessary noise right now. After doing so, I got pretty far into the setup process before hitting a MySQL error.

I had to do the following things, some of which were code changes to the `main` branch, some of which were undocumented in the README but I was able to put together with a bit of guesswork. All the changes I made to the code in `main` are documented in `main-notraefik`, so it should be possible to reproduce by checking out `main-notraefik` and proceeding through the README instructions, adding steps 1 and 3 here. (Step 2 is a change to the code, so it's already done for `main-notraefik`.)

1. I hadn't cloned the repo using the `--recursive` flag, so I had to do `git submodule update --init` to get the BookwormDB and Vega repos.
2. I had to modify `compose/bookworm/Dockerfile` because it was trying to `cd` into a nonexistent directory.
3. I had to modify the `bookworm.cnf` file generated by this command (given in the README) `echo "[client]\ndatabase=test_bookworm_files" >> corpus/bookworm.cnf`. One issue was that the command leaves the `\n` as a literal value rather than inserting a newline. (I am not good at bash quote conventions so I don't know offhand why or how to fix.) But also, the build command wanted additional info about the database, which I was able to guess. So in my `/corpus` dir, `bookworm.cfg` now reads:

        [client]
        database=test_bookworm_files
        host=mariadb  # in case this wasn't already clear, service names are URLs in the internal network
        user=root
        password=insecure_root_dev_password

I can then run `bin/run build_bookworm` and it works for a moment, before giving the following error:

```
[CreateDatabase.py:111-query() 07:28:29.986] Query failed: 
INSERT INTO tmp SELECT bookid,nwords,title__id,author__id,fedNumber__id,paragraphNumber__id,date_year,date_month,date_day,date_week_year FROM catalog USE INDEX () LEFT JOIN title__id USING (title )  LEFT JOIN author__id USING (author )  LEFT JOIN fedNumber__id USING (fedNumber )  LEFT JOIN paragraphNumber__id USING (paragraphNumber ) 

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/CreateDatabase.py", line 98, in query
    cursor.execute(sql)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 206, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 319, in _query
    db.query(q)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/connections.py", line 259, in query
    _mysql.connection.query(self, query)
MySQLdb._exceptions.OperationalError: (1048, "Column 'nwords' cannot be null")

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/bookworm", line 8, in <module>
    sys.exit(run_arguments())
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/manager.py", line 614, in run_arguments
    getattr(my_bookworm,args.action)(args)
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/manager.py", line 204, in build
    self.prep(args)
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/manager.py", line 216, in prep
    getattr(self, args.goal)(args)
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/manager.py", line 270, in all
    self.database_metadata(args)
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/manager.py", line 357, in database_metadata
    Bookworm.create_fastcat_and_wordsheap_disk_tables()
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/CreateDatabase.py", line 505, in create_fastcat_and_wordsheap_disk_tables
    self.db.query(q)
  File "/usr/local/lib/python3.8/site-packages/bookwormDB/CreateDatabase.py", line 107, in query
    cursor.execute(sql)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 206, in execute
    res = self._query(query)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 319, in _query
    db.query(q)
  File "/usr/local/lib/python3.8/site-packages/MySQLdb/connections.py", line 259, in query
    _mysql.connection.query(self, query)
MySQLdb._exceptions.OperationalError: (1048, "Column 'nwords' cannot be null")
done
```

That sounds to me like a DB schema constraint violation of some kind? But I am not sure and am going to stop with the guesswork here and say — any idea what might fix this @bmschmidt?

**Comment by bmschmidt** [comment]
Hmm, I'll take a look. Not sure if I've seen null happening that way--the logical inference there would be that some documents have zero words in them, which doesn't make much sense. Will see if I'm able to tear down my local install a little further to reproduce.

On the bookworm.cnf file: it's correct to only have the name of the database in there. Host, username, and password should 
all be read from /etc/bookworm/admin.cnf. Not sure why that wouldn't be there, but I'm going to set up a different way of populating that file right now.
```
 [client]
 database=test_bookworm_files
```

Ugh, git submodules kind of suck, but you can the script in `bin/run` that I use to pip install -e bookworm to make live-sync edits to the running container; that's very useful to me for now.

**Comment by bmschmidt** [comment]
Unable to tell what's the issue, but I just cleaned up the build command a bit more. This still works for me, and I've changed the instructions to describe the new process, which is a little simpler and better with the passwords. (E.g., it no longer contains the mysql passwords in the images, and just inserts them through runtime variables.)

I've updated my command at `bin/run rebuild` to include this:

```
docker-compose down --volumes --rmi all --remove-orphans && docker volume rm mysql_data && docker volume create --name=mysql_data
```

So I'm not sure how it could be broken on yours but work for me. If it doesn't with the bare-bones bookworm.cnf, let me know?
